File: demo_oneactivity_multifragment/src/main/java/com/example/demo_oneactivity_multifragment/BaseFragment.java
package com.example.demo_oneactivity_multifragment;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

public abstract class BaseFragment extends Fragment {
    public View rootView;
    public Activity mActivity;
    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        this.mActivity = activity;
    }

    @Nullable
    @Override
    public final View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        if (rootView == null) {
            rootView = inflater.inflate(this.getLayoutId(), container, false);
        }
        return rootView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        init(savedInstanceState);
    }

    /**
     * fragment 返回事件
     */
    public void goBack(){
        mActivity.onBackPressed();
    }
    protected abstract int getLayoutId();

    protected abstract void init(Bundle savedInstanceState);
}

File: demo_kotlin/src/main/java/com/practice/eyepetizer/base/BaseFragment.kt
package com.tt.lvruheng.eyepetizer.ui.fragment

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

abstract class BaseFragment : Fragment() {
    var mActivity: Context? = null
    var rootView: View? = null
    var isViewInitiated: Boolean = false
    var isDataLoaded: Boolean = false
    override fun onAttach(context: Context?) {
        super.onAttach(context)
        mActivity = context
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        if (rootView == null) {
            rootView = inflater?.inflate(getLayoutId(), container, false)
        }
        return rootView
    }


    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        isViewInitiated = true
        init(savedInstanceState)
        prepareFetchData()
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        prepareFetchData()
    }

    private fun prepareFetchData() : Boolean {
        return prepareFetchData(false)
    }

    private fun prepareFetchData(forceUpdate: Boolean): Boolean {
        if(userVisibleHint && isViewInitiated &&(!isDataLoaded || forceUpdate)){
            loadData()
            isDataLoaded = true
            return true
        }
        return false
    }

    /*子类实现*/
    abstract fun getLayoutId(): Int

    abstract fun loadData()

    abstract fun init(savedInstanceState: Bundle?)


}